traffic_ops/traffic_ops_golang/deliveryservice/health.go

```diff
@@ -0,0 +1,142 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"errors"
+	"net/http"
+	"strings"
+
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/tenant"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/util/monitorhlp"
+)
+
+func GetHealth(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"id"}, []string{"id"})
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	dsID := inf.IntParams["id"]
+
+	userErr, sysErr, errCode = tenant.CheckID(inf.Tx.Tx, inf.User, dsID)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	ds, cdn, ok, err := dbhelpers.GetDSNameAndCDNFromID(inf.Tx.Tx, dsID)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting delivery service name from ID: "+err.Error()))
+		return
+	}
+	if !ok {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusNotFound, nil, nil)
+	}
+
+	health, err := getHealth(inf.Tx.Tx, ds, cdn)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting delivery service health: "+err.Error()))
+		return
+	}
+
+	api.WriteResp(w, r, health)
+}
+
+func getHealth(tx *sql.Tx, ds tc.DeliveryServiceName, cdn tc.CDNName) (tc.HealthData, error) {
+	monitors, err := monitorhlp.GetURLs(tx)
+	if err != nil {
+		return tc.HealthData{}, errors.New("getting monitors: " + err.Error())
+	}
+	monitor, ok := monitors[cdn]
+	if !ok {
+		return tc.HealthData{}, nil // TODO emulates old Perl behavior; change to return error?
+	}
+	return getMonitorHealth(tx, ds, monitor)
+}
+
+func getMonitorHealth(tx *sql.Tx, ds tc.DeliveryServiceName, monitorFQDN string) (tc.HealthData, error) {
+	client, err := monitorhlp.GetClient(tx)
+	if err != nil {
+		return tc.HealthData{}, errors.New("getting monitor client: " + err.Error())
+	}
+
+	totalOnline := uint64(0)
+	totalOffline := uint64(0)
+	cgData := map[tc.CacheGroupName]tc.HealthDataCacheGroup{}
+
+	crStates, err := monitorhlp.GetCRStates(monitorFQDN, client)
+	// TODO on err, try another online monitor
+	if err != nil {
+		return tc.HealthData{}, errors.New("getting CRStates for delivery service '" + string(ds) + "' monitor '" + monitorFQDN + "': " + err.Error())
```

Comment:
nit but these could probably leverage `fmt.Errorf` to be more concise

Reply:
So, I try to avoid Go format strings, because they're 10-100x slower than string concatenation. We hit that issue with Grove, and it was the difference in maxing out at 2gbps, and being able to saturate a 20g NIC

I also hugely prefer format strings, I always use them in C/C++. And yeah, it's a fine line between Premature Optimization. Odds are TO will never need to do that level of performance. But at the same time, I feel like it's a good habit to just always avoid them, and then it's never a problem